mdagent/tools/base_tools/preprocess_tools/uniprot.py

```diff
@@ -0,0 +1,1250 @@
+import time
+from enum import Enum
+
+import requests
+from langchain.tools import BaseTool
+from requests.adapters import HTTPAdapter
+from urllib3.util import Retry
+
+
+class SiteType(Enum):
+    ACTIVE = ("ft_act_site", "active site")
+    BINDING = ("ft_binding", "binding site")
+    SITES = ("ft_site", "site")
+
+
+class PTMType(Enum):
+    CHAIN = ("ft_chain", "Chain")
+    CROSSLINK = ("ft_crosslnk", "Cross-link")
+    DISULFIDE_BOND = ("ft_disulfid", "Disulfide bond")
+    GLYCOSYLATION = ("ft_carbohyd", "Glycosylation")
+    INITIATOR_METHIONINE = ("ft_init_met", "Initiator methionine")
+    LIPIDATION = ("ft_lipid", "Lipidation")
+    MODIFIED_RESIDUE = ("ft_mod_res", "Modified residue")
+    PEPTIDE = ("ft_peptide", "Peptide")
+    PROPEPTIDE = ("ft_propep", "Propeptide")
+    SIGNAL_PEPTIDE = ("ft_signal", "Signal peptide")
+    TRANSIT_PEPTIDE = ("ft_transit", "Transit peptide")
+
+
+class StructureMap(Enum):
+    BETA = ("ft_strand", "Beta strand")
+    HELIX = ("ft_helix", "Helix")
+    TURN = ("ft_turn", "Turn")
+
+
+class QueryUniprot:
+    API_URL = "https://rest.uniprot.org"
+
+    def get_sequence_mapping(
+        self,
+        query: str,
+        from_db: str = "UniProtKB_AC-ID",
+        to_db: str = "PDB",
+        polling_interval: int = 3,
+    ) -> list:
+        """
+        Fetch specific ID mapping from UniProt and extract the 'to' field f
+        rom results.
+
+        Args:
+            query: The UniProt ID to map (e.g. 'P05067')
+            from_db: The source database to map from.
+                Defaults to 'UniProtKB_AC-ID'.
+            to_db: The target database to map to. Defaults to 'PDB'.
+            polling_interval: The interval to poll the API for results.
+                Defaults to 3 seconds.
+
+        Returns:
+            A list of mapped database entries from the 'to' field if
+                successful, otherwise an empty list.
+        """
+        with requests.Session() as session:
+            session.mount(
+                "https://",
+                HTTPAdapter(
+                    max_retries=Retry(
+                        total=5,
+                        backoff_factor=0.25,
+                        status_forcelist=[500, 502, 503, 504],
+                    )
+                ),
+            )
+            response = session.post(
+                f"{self.API_URL}/idmapping/run",
+                data={"from": from_db, "to": to_db, "ids": query},
+            )
+            response.raise_for_status()
```

Comment:
Does it need handlers in case of exceptions? Or are they handle internally, and just keeps retrying?

Reply:
fixed